Feature/start with last opened tabs

PoisonousJohn opened the pull request:
- Added "Continue where you left off" setting to the settings page.
- Added "Session" entity to save session-related data to separate file
- Added list of tab states to session.
- Implemented restoration of tabs when opening first browser window

Commit:
Finished main work on #50

Commit:
Finished main work on #50

PoisonousJohn commented:
This is for #50

timsueberkrueb commented:
@JohnPoison thank you! I will hopefully be able to look into it tomorrow (or on Friday).

PoisonousJohn commented:
Also I'm currently looking into saving tab's scrolling progress upon exit. It seems that for this task we'll have to use WebChannel. Also this will give us more flexible way to communicate QML <--> WebPage content

timsueberkrueb commented:
@JohnPoison That would be cool :) If we have to use WebChannels, we should merge `feature/qtwebengine` first. But couldn't we just get the scroll position by running some JavaScript in the page as we do for theme color support?

PoisonousJohn commented:
@tim-sueberkrueb sure we can. But isn't that lot easier to modify property of a js object, instead of messing with runJavaScript callbacks? Besides it's not so easy to use runJavaScript callback, since I'm saving sesison data on window close. So I should block closing until callback returns.

I see that web channel can give us a handy way to communicate with web page content. Well, we can even do this in a separate PR

timsueberkrueb commented:
@JohnPoison I see, that's right. We could also use WebChannels for theme color handling then.

timsueberkrueb commented:
`feature/qtwebengine` was merged with develop (including the travis fix). You'll probably want to rebase your work on develop. Thanks!

PoisonousJohn commented:
Lol, it seems i've messed up with git. Will fix it 10th of May.

PoisonousJohn commented:
@tim-sueberkrueb rebase done.

Commit:
Added missing copyrights header

Commit:
Added missing copyrights header

timsueberkrueb commented on src/ui/tabview/content/SettingsContent.qml:
```diff
@@ -61,11 +61,35 @@ TabContent {
 
                 TitleLabel {
-                    text: "Start"
+                    text: "Startup"
```
Please use `qsTr()` now

timsueberkrueb commented on src/ui/tabview/content/SettingsContent.qml:
```diff
@@ -61,11 +61,35 @@ TabContent {
+                }
+
+                Column {
+                    Label { text: "Start new window with" }
```
Please use `qsTr()` now

timsueberkrueb reviewed (changes requested):
Hey, I'm so sorry that I forgot about this PR ._.
Can you rebase it on `develop`?
Looks good otherwise :+1:
We might want to rethink settings (in general) though (but that's something for v0.5.0 ;)).

timsueberkrueb commented on src/core/session/session.cpp:
```diff
@@ -0,0 +1,116 @@
+ * $END_LICENSE$
+*/
+
+#include "Session.h"
```
I think that's `"session.h"`.

timsueberkrueb commented:
I think one thing to consider is whether it would be better if restored tabs would only load on demand (e.g. when the user activates them) as it is the case in Firefox. That would prevent long loading times when many tabs are stored.
What do you think?

ShalokShalom commented:
Completely agree. +1 

Just think about cases with 20-40 tabs open and more..

timsueberkrueb commented:
Another problem with this I haven't thought of yet is multi window support.

Commit:
fixed sesson header

Commit:
do not load all tabs upon restoring

PoisonousJohn commented:
Rebased on develop.
Added lazy loading of the tabs.

Regarding multi-window support, right now the session is being saved for the last closed window. I think if we need multi-window support, it should be done as a separate issue.

timsueberkrueb commented on src/ui/tabview/TabController.qml:
```diff
@@ -50,11 +50,24 @@ QtObject {
+    {
+        for (var i = 0; i < tabContentView.pages.count; ++i)
+        {
+            var tabPage = tabContentView.pages.get(i).item;
```
Let's stay consistent with the rest of the JavaScript code and keep the opening curly brace in the same line ;)
```js
for (var i = 0; i < tabContentView.pages.count; ++i) {
```

timsueberkrueb commented on src/ui/tabview/TabController.qml:
```diff
@@ -50,11 +50,24 @@ QtObject {
-            background: background
-        });
+    function completeTabsRestore()
+    {
```
Let's keep the opening curly brace in the same line for consistency ;)
```javascript
function completeTabsRestore() {
```

timsueberkrueb commented on src/ui/tabview/TabPage.qml:
```diff
@@ -75,24 +84,38 @@ Item {
-                );
-                break;
+        if (!loadContent)
+        {
```
Let's keep the opening curly brace in the same line for consistency.

timsueberkrueb commented on src/ui/tabview/TabPage.qml:
```diff
@@ -75,24 +84,38 @@ Item {
+            newContent = lazyWebContentComponent.createObject(contentContainer, lazyProperties)
         }
+        else
+        {
```
Let's keep the opening curly brace in the same line for consistency.

timsueberkrueb commented on src/ui/tabview/content/LazyWebContent.qml:
```diff
@@ -0,0 +1,65 @@
+        id: lazyCreateConnection
+
+        function tryCreateContent()
+        {
```
Let's keep the opening curly brace in the same line for consistency.

timsueberkrueb commented on src/ui/tabview/content/LazyWebContent.qml:
```diff
@@ -0,0 +1,65 @@
+                return;
+
+            if (parentTabPage.loadContent)
+            {
```
Let's keep the opening curly brace in the same line for consistency.

timsueberkrueb commented on src/core/session/session.cpp:
```diff
@@ -87,8 +89,16 @@ void Session::load()
+
+    m_activeTab = std::max(m_tabs.count() - 1, 0);
+    if (root.find("activeTab") != root.end())
+    {
```
Let's keep the opening curly brace in the same line for consistency-

PoisonousJohn replied:
sorry for that braces, my old habit from c# :)

timsueberkrueb replied:
No problem, thanks! :)

timsueberkrueb commented on src/core/session/session.h:
```diff
@@ -0,0 +1,58 @@
+    Q_PROPERTY(int activeTab READ getActiveTab)
+
+    int getActiveTab() { return m_activeTab; }
+signals:
```
We can remove the `signals` and `public slots` sections as they are not used.

PoisonousJohn replied:
sure

timsueberkrueb reviewed (commented):
Thanks a lot for getting back to this!
Looking good, will check it out and test myself asap.

timsueberkrueb reviewed (commented):
Looks like the data directory (`/home/tim/.local/share/liri-browser` in my case) is not being created:
`Couldn't open session file for write!`

We should make sure the directory gets created, see https://github.com/lirios/browser/blob/develop/src/core/settings/settings.cpp#L40

timsueberkrueb reviewed (commented):
One bug I noticed is that it doesn't restore `liri://` tabs. For example, when restarting the browser having a `liri://settings` tab opened, it won't reload. A `This site can’t be reached` page shows up instead.

timsueberkrueb commented:
Thinking about it, couldn't we merge `LazyWebContent` and `WebContent`? I'm thinking about making `WebContent` just contain the loader wrapping the `WebEngineView` and getting an additional `lazy` property to toggle default or lazy loading.

timsueberkrueb commented on src/core/session/tabstate.cpp:
```diff
@@ -0,0 +1,73 @@
+
+float TabState::readingProgress() const
+{
+    return m_readingProgress;
```
Should we remove readingProgress while it's not being used, yet?

PoisonousJohn replied:
Yep, sure

timsueberkrueb commented on src/core/session/session.cpp:
```diff
@@ -0,0 +1,130 @@
+        tabObject["url"] = tab->url().toString();
+        tabObject["title"] = tab->title();
+        tabObject["icon"] = tab->iconUrl().toString().replace("image://favicon/","");
+        tabObject["readingProgress"] = 0.f;
```
The JSON fields for settings are currently all `snake_case`. I do prefer `camelCase` though. We can change the settings schema to use `camelCase` (which shouldn't be a problem right now as we don't have a release yet).
Any objections?

PoisonousJohn replied:
No objections. Should I leave it like that for now?

tim-sueberkrueb replied:
Yep.

timsueberkrueb replied:
Yep, let's :)

PoisonousJohn commented:
> Thinking about it, couldn't we merge LazyWebContent and WebContent? I'm thinking about making WebContent just contain the loader wrapping the WebEngineView and getting an additional lazy property to toggle default or lazy loading.

I have thought about it, and even tried, but something went wrong. Let me try again.
Then it should fix problems with the settings page as well

PoisonousJohn commented:
> Looks like the data directory (/home/tim/.local/share/liri-browser in my case) is not being created:
> Couldn't open session file for write!
> We should make sure the directory gets created, see https://github.com/lirios/browser/blob/develop/src/core/settings/settings.cpp#L40

I almost sure that I have fixed it. It seems I've lost the change during problems with rebase. Let me fix it again

ShalokShalom commented:
Oh rebase ^-^

ShalokShalom commented:
https://hackernoon.com/dont-fear-the-rebase-bca683888dae

Commit:
fixes according to the review

Commit:
forgotten stylefix

Commit:
more stylefixes

PoisonousJohn commented:
This iteration looks more appropriate. liri-settings bug is fixed as well

timsueberkrueb commented:
Awesome, thank you! Will check it out a bit later today!

timsueberkrueb commented on src/ui/tabview/TabPage.qml:
```diff
@@ -59,6 +55,15 @@ Item {
     property int tabType
 
+    function tryCreateContent() {
+        console.log("Try create content");
```
Do you want to keep this logging?

timsueberkrueb commented on src/core/session/session.cpp:
```diff
@@ -42,6 +42,11 @@ Session::Session(QObject *parent)
 void Session::save(TabsModel* tabs)
 {
+    if (!QDir(Paths::DataLocation).exists()) {
+        qDebug() << "DataLocation path doesn't exist.";
```
Let's make this a bit more human-friendly, like "App data path doesn't exist".

timsueberkrueb commented on src/ui/tabview/TabPage.qml:
```diff
@@ -69,56 +74,82 @@ Item {
     }
 
+    onActiveTabChanged: {
+        console.log("onActiveTabChanged");
```
Can we remove this debug statement?

timsueberkrueb commented on src/core/session/session.cpp:
```diff
@@ -0,0 +1,133 @@
+{
+    QVariantList tabs;
+    for (TabState* state : m_tabs)
+    {
```
Found another curly brace :)

timsueberkrueb commented on src/ui/tabview/TabPage.qml:
```diff
@@ -27,11 +27,16 @@ import ".."
 Item {
     id: tabPage
 
+    // binding. should be set from outside
```
Is this comment still relevant? I don't understand what it means.

PoisonousJohn replied:
It's relevant. It means that activeTab property itself doesn't do anything. It should be a functions assigned by another component. This function should return an activeTab. I did this to avoid dependency on TabController

timsueberkrueb replied:
Ah, all right.

timsueberkrueb commented on src/ui/tabview/TabPage.qml:
```diff
@@ -50,6 +55,15 @@ Item {
+        if (contentItem || !loadContent || tab !== activeTab)
+            return;
+
+        console.log("Creating content");
```
Do we need this debug statement?

timsueberkrueb reviewed (changes requested):
Works great now, thank you! One thing I noticed is that you store the icon url but the icons will only show up when the tab fully loads. We either should make the icons load or drop the property.
We also probably want to prevent saving the icon url for `liri://` tabs:
```json
"icon": "qrc:/liri.io/imports/Fluid/Controls/icons/action/settings.svg",
```
It also looks like the icon information gets lost when I restart the browser, tabs get restored but I don't activate them and restart the browser again:
```json
"icon": ""
```
Besides of that, looking good!

timsueberkrueb commented on src/core/session/session.cpp:
```diff
@@ -0,0 +1,133 @@
+        tabObject["url"] = tab->url().toString();
+        tabObject["title"] = tab->title();
+        tabObject["icon"] = tab->iconUrl().toString().replace("image://favicon/","");
+        tabObject["readingProgress"] = 0.f;
```
`readingProgress` can be removed here, right?

PoisonousJohn commented:
> We also probably want to prevent saving the icon url for liri://
I would leave it as is for consistency, since it works fine. I don't want to check url for "liri://".

PoisonousJohn commented on src/ui/window/BrowserWindow.qml:
```diff
@@ -502,7 +502,7 @@ ApplicationWindow {
                     'url': restoreTabs[i].url,
                     'title': restoreTabs[i].title,
-                    'iconUrl': restoreTabs[i].iconUrl
+                    'iconUrl': restoreTabs[i].icon.replace("https", "http")
```
This bothers me. Right now QML Image is not able to load images via https, showing SSL error in console. Yet browser works fine with SSL. Don't know how to fix it, so it's a hack right now.

timsueberkrueb replied:
Interesting, I cannot reproduce this issue here on Ubuntu 17.10 with Qt 5.10.1.

Commit:
fixed loading of icons, few style fixes as well

timsueberkrueb commented on src/ui/tabview/content/SettingsContent.qml:
```diff
@@ -61,11 +61,35 @@ TabContent {
-                Label {
-                    text: qsTr("Start url")
+                Column {
+                    Label { text: "Start new window with" }
```
Use `qsTr()`

timsueberkrueb commented:
Looking good otherwise!

Commit:
added missing qsTr

PoisonousJohn commented:
@tim-sueberkrueb  I believe everything is ok now :)

timsueberkrueb commented:
@PoisonousJohn I think you need to remove `replace("https", "http")` :)

PoisonousJohn commented:
@tim-sueberkrueb yeah, sure, I forgot about that, still didn't test it in a "Empty" project

Commit:
removed https hack

PoisonousJohn commented:
@tim-sueberkrueb done

timsueberkrueb commented:
@PoisonousJohn do you have any preference on merging the PR? Should I squash your commits?

timsueberkrueb commented:
Thank you @PoisonousJohn!
